// @meteorwallet/sdk/dist/meteor-common/src/modules_external/near_public_indexer/types/near_indexer_basic_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ENearIndexer_ActionType = exports.ENearIndexer_AccessKeyPermission = void 0;
var ENearIndexer_AccessKeyPermission;
(function (ENearIndexer_AccessKeyPermission) {
    ENearIndexer_AccessKeyPermission["FULL_ACCESS"] = "FULL_ACCESS";
    ENearIndexer_AccessKeyPermission["FUNCTION_CALL"] = "FUNCTION_CALL";
})(ENearIndexer_AccessKeyPermission = exports.ENearIndexer_AccessKeyPermission || (exports.ENearIndexer_AccessKeyPermission = {}));
var ENearIndexer_ActionType;
(function (ENearIndexer_ActionType) {
    ENearIndexer_ActionType["TRANSFER"] = "TRANSFER";
    ENearIndexer_ActionType["FUNCTION_CALL"] = "FUNCTION_CALL";
    ENearIndexer_ActionType["DELETE_KEY"] = "DELETE_KEY";
    ENearIndexer_ActionType["DEPLOY_CONTRACT"] = "DEPLOY_CONTRACT";
    ENearIndexer_ActionType["CREATE_ACCOUNT"] = "CREATE_ACCOUNT";
    ENearIndexer_ActionType["DELETE_ACCOUNT"] = "DELETE_ACCOUNT";
})(ENearIndexer_ActionType = exports.ENearIndexer_ActionType || (exports.ENearIndexer_ActionType = {}));
