// @meteorwallet/sdk/dist/meteor-common/src/modules_feature/missions/feature_enrollment_records/feature_enrollment_records.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EFeatureEnrollmentRecord_KeyStatus = exports.EFeatureEnrollmentRecord_ConsentStatus = exports.EFeatureEnrollmentRecord_EnrollableFeatureId = void 0;
var EFeatureEnrollmentRecord_EnrollableFeatureId;
(function (EFeatureEnrollmentRecord_EnrollableFeatureId) {
    EFeatureEnrollmentRecord_EnrollableFeatureId["mission"] = "mission";
})(EFeatureEnrollmentRecord_EnrollableFeatureId = exports.EFeatureEnrollmentRecord_EnrollableFeatureId || (exports.EFeatureEnrollmentRecord_EnrollableFeatureId = {}));
var EFeatureEnrollmentRecord_ConsentStatus;
(function (EFeatureEnrollmentRecord_ConsentStatus) {
    EFeatureEnrollmentRecord_ConsentStatus["pending"] = "pending";
    EFeatureEnrollmentRecord_ConsentStatus["accepted"] = "accepted";
    EFeatureEnrollmentRecord_ConsentStatus["revoked"] = "revoked";
    EFeatureEnrollmentRecord_ConsentStatus["denied"] = "denied";
})(EFeatureEnrollmentRecord_ConsentStatus = exports.EFeatureEnrollmentRecord_ConsentStatus || (exports.EFeatureEnrollmentRecord_ConsentStatus = {}));
var EFeatureEnrollmentRecord_KeyStatus;
(function (EFeatureEnrollmentRecord_KeyStatus) {
    EFeatureEnrollmentRecord_KeyStatus["inactive"] = "inactive";
    EFeatureEnrollmentRecord_KeyStatus["active"] = "active";
    EFeatureEnrollmentRecord_KeyStatus["revoked"] = "revoked";
    EFeatureEnrollmentRecord_KeyStatus["low_allowance"] = "low_allowance";
    EFeatureEnrollmentRecord_KeyStatus["wrongly_configured"] = "wrongly_configured";
})(EFeatureEnrollmentRecord_KeyStatus = exports.EFeatureEnrollmentRecord_KeyStatus || (exports.EFeatureEnrollmentRecord_KeyStatus = {}));
